### ISSUE TITLE
Is the code in this repository only for inference?

### ISSUE THREAD

**Issue by CodeTal** [facebookresearch/llama#961]
Can we finetune a llama using the model structure defined in this repository? 

I know we can use Huggingface codes to do the finetune. But I want to slightly modify the model architecture then do the finetune. The Huggingface class seems not flexible enough to do that. I have tried to use these code to finetune (build a Transformer class, load checkpoints, then use the Transformer to update the weights), but a lot of bug occurs.

**Comment by subramen** [comment]
You should be able to directly finetune llama models from this repo, but we don't have example scripts for that. The benefit with using HF of course is it abstracts a lot of the boilerplate. @osanseviero might have some insights into subclassing the Llama model

**Comment by CodeTal** [comment]
> You should be able to directly finetune llama models from this repo, but we don't have example scripts for that. The benefit with using HF of course is it abstracts a lot of the boilerplate. @osanseviero might have some insights into subclassing the Llama model

Thanks for your reply. However, I encountered problems during finetuning. I tried to freeze all the parameters of the model except for the last linear output layer to do fine-tune. After the first step of the optimizer, the weights of the last output layer became NaN. I have tried low learning rates and different optimizers, and I have examined the gradients and the logits of the model, which are not NaN.

It should be a common problem and easy to replicate. But it is difficult for me to identify the root of this problem. For your reference, here are some of my codes:

1. I commented the @torch.inference_mode() line of the forward() function of the Transformer class in llama/model.py.

2. I use this code to freeze most of the parameters of the model:

# Freeze the LM
for param in model.parameters():
    param.requires_grad = False
model.output.weight.requires_grad = True

3. I use this code to do the training of first iteration (after this iteration the weights of the output layer became NaN):

optimizer.zero_grad()
logits = model(tokens[:, prev_pos:min_prompt_len], prev_pos)
probs = torch.softmax(logits[:, -1] / temperature, dim=-1)
loss = cross_entropy_loss(probs, tokenized_ans.view(-1)).to(device)
loss.backward()
optimizer.step()

That's a simple implementation, but I have spent many days fixing on various problems. Therefore I was wondering if the code in this repository cannot be directly used for finetuning.

**Comment by subramen** [comment]
You can try debugging by checking what the output layer's weights and gradients are. Are they of the same dtype and on the same device? If you can share a minimal reproducible example, I'd be happy to take a look

**Comment by subramen** [comment]
Closing the issue, feel free to reopen when you have more information

**Comment by waterluck** [comment]
@CodeTal have you solved this problem? I also want to freeze a few parameters of the model and then fine-tune it, but in many layers, the parameter's torch size did not update with training. Have you met the same problem or had any solution on it?
